### DISCUSSION TITLE
Proposal: Support Async Pattern Methods (Cω Chords)

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#88] [General]
Copied from https://github.com/dotnet/roslyn/issues/1165

## Background

In my opinion one of the more interesting ideas in the Polyphonic C# and Cω languages out of Microsoft Research was "asynchronous method chords".  Combined with asynchronous methods this simple syntax provided one of the most elegant ways to express complex synchronization between two or more asynchronous operations.
## Proposal

An asynchronous pattern is comprised of two parts.
#### Mailbox Methods

The first part is an asynchronous method which has no body and must return `void`.  This method serves as a notification primitive.  Invoking the method is the same as posting a message where the arguments are additional pieces of data transmitted with the message.

``` cs
public async void Add(T value);
public async void Done();
```

When invoked a mailbox method effectively returns immediately to the caller and can never throw exceptions.
#### Pattern Methods

The second part is an asynchronous pattern method which may return `void`, `Task` or `Task<T>` and has a normal method body.  This method is decorated with one or more `&` join operators with the name and signature of a mailbox method between the parameter list and the method body.  The parameter names in the joined mailbox signature may be renamed.  The values of those parameters are available in the body of the pattern method.

``` cs
public async Task<T> Take()
    & Add(T value)
{
    return value;
}
```

After the pattern method body more join operations may appear using different permutations of the available mailbox methods.  These permutations must be unique per each method body regardless of order.  When the pattern method is invoked which body is executed depends on which mailbox methods are invoked.

``` cs
public async Task<T> Take()
    & Add(T value)
{
    return value;
}
    & Done()
{
   throw new TaskCanceledException("No more items!");
}
```

What happens when the pattern method depends on whether there have already been corresponding calls to the mailbox methods.  For a pattern method body to be executed there must be corresponding calls to every mailbox method on which it joins.

If there are no corresponding calls to the mailbox methods for any of the pattern methods then the method returns immediately.  Once a mailbox method is called which satisfies the pattern method then that pattern method body will be executed on the same thread that invoked the mailbox method.

If the method returns `Task` or `Task<T>` the returned task is incomplete.  If there are corresponding calls to the signal methods for a pattern method then the method is invoked on the current thread.

One the pattern method body is executing it behaves like a normal asynchronous method.  It may use the `await` keyword to await on other asynchronous operations, including other pattern methods.  It may return a value which sets the result on the returned `Task` or `Task<T>` instance, or it may throw an exception which sets the exception on the returned `Task` or `Task<T>` instance.

``` cs
public class PatternQueue<T>
{
    public async void Add(T value);
    public async void Done();

    public async Task<T> Get()
        & Add(T value)
    {
        return value;
    }
        & Done()
    {
        throw new TaskCanceledException();
    }
}

static class Program
{
    static void Main()
    {
        var queue = new PatternQueue<int>();

        var task1 = queue.Get();
        Debug.Assert(task1.IsCompleted == false);

        queue.Add(5);
        Debug.Assert(task1.IsCompleted == true);
        Debug.Assert(task1.Result == 5);

        queue.Add(6);
        queue.Add(7);
        var task2 = queue.Get();
        Debug.Assert(task2.IsCompleted == true);
        Debug.Assert(task2.Result == 6);

        queue.Done();
        var task3 = queue.Get();
        Debug.Assert(task3.IsCompleted == true);
        Debug.Assert(task3.IsCanceled == false);
        Debug.Assert(task3.Result == 7);

        var task4 = queue.Get();
        Debug.Assert(task4.IsCanceled == true);
    }
}
```

----

### Reference material:

[Cω Overview](http://research.microsoft.com/en-us/um/cambridge/projects/comega/doc/comega_whatis.htm)
[Cω Concurrency Examples Tutorial](http://research.microsoft.com/en-us/um/cambridge/projects/comega/doc/comega_tutorials_concurrency_extensions.htm)
[Cω Concurrency Specification](http://research.microsoft.com/en-us/um/cambridge/projects/comega/doc/comega_langspec_concurrencyconstructs.htm)

Note that I am only using the Cω as a reference.  The syntax that I proposed is different in that it builds on `async/await` from C# 5.0.  From a functionality point of view the biggest difference is that the pattern method is not synchronous, as it is in Cω.  If the pattern method has no corresponding signal method invocations it will not block but rather return an incomplete `Task`.

**Comment by jnm2** [comment]
Third code block has `Task<T> Task()` instead of `Task<T> Take()`.
Interesting! My first thought is, why can't this be done as a library?

**Comment by HaloFour** [comment]
@jnm2 

> Third code block has `Task<T> Task()` instead of `Task<T> Take()`.

Fixed.

> Interesting! My first thought is, why can't this be done as a library?

It can be, although it's not as nice.  I wrote this prior to `async`/`await`:

http://joins.codeplex.com/
